generate_latest_json.py

```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env python3
+"""
+Sopel update-check JSON generation utility
+
+This script creates a `latest.json` file representing the most recent stable and
+unstable `sopel` releases published to PyPI, using PyPI's own JSON API.
+
+Copyright 2024 dgw, technobabbl.es
+Licensed under the Eiffel Forum License 2.
+
+https://sopel.chat
+"""
+import json
+
+from packaging.version import parse as parse_version
+import requests
+
+
+def main(argv=None):
+    data = requests.get('https://pypi.org/pypi/sopel/json').json()
+    releases = sorted([parse_version(v) for v in data['releases'].keys()])
+
+    latest = releases[-1]
+    stable = latest
+    if latest.is_prerelease:
+        for v in reversed(releases[:-1]):
+            if not v.is_prerelease:
+                stable = v
+                break
+        else:
+            print("ERROR: Latest version is a prerelease, but no earlier stable version found!")
+            exit(1)
+
+    with open('latest.json', 'w') as f:
+        f.write("{{\n"  # double braces so Python doesn't get confused
+                "    \"version\": \"{stable}\",\n"
+                "    \"unstable\": \"{latest}\",\n"
+                "    \"release_notes\": \"https://sopel.chat/changelog/{stable}/\",\n"
+                "    \"unstable_notes\": \"https://github.com/sopel-irc/sopel/releases/v{latest}\"\n"
+                "}}".format(latest=latest, stable=stable))
```

Comment:
I was originally going to suggest swapping the outer `""` for `''` here, but it's probably even better to use `json` here, I find it a lot easier to read personally.

I did invoke the bête noire of f-strings, but feel free to change those to `.format()` if you agree with the broad strokes but prefer not to use the interpolated literal.

```suggestion
        sopel_version_info = {
                "version": str(stable),
                "unstable": str(latest),
                "release_notes": f"https://sopel.chat/changelog/{stable}",
                "unstable_notes": f"https://github.com/sopel-irc/sopel/releases/v{latest}",
        }
        json.dump(sopel_version_info, f)
```

Reply:
Since `stable` and `latest` both need to be used in multiple places, I actually don't fancy trying to make that look nice with `.format()`. f-strings are grudgingly acceptable here, where there are no l10n concerns :P